fix(anthropic): Async and streaming Anthropic  (#302)

AromaticHydrocarbon opened the pull request:
<!-- Thanks for submitting a PR! To make sure this gets merged quickly, make sure to check the following checkboxes. -->

- [x] I have added tests that cover my changes.
- [x] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [x] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

![image](https://github.com/traceloop/openllmetry/assets/71716390/9fd63e31-6178-40d4-baef-a041ad47da99)
![image](https://github.com/traceloop/openllmetry/assets/71716390/38942363-dd88-4679-b081-c74d56600ddd)
![image](https://github.com/traceloop/openllmetry/assets/71716390/87504926-cc6c-4540-ad63-5cffed7fcaee)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/traceloop/openllmetry?pullRequest=750) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/traceloop/openllmetry?pullRequest=750) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/traceloop/openllmetry?pullRequest=750) it.</sub>

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/__init__.py:
```diff
@@ -178,47 +254,213 @@ def wrapper(wrapped, instance, args, kwargs):
-                    _set_response_attributes(span, response)
-                    _set_token_usage(span, instance._client, kwargs, response)
+@_with_tracer_wrapper
+async def _a_wrap(tracer, to_wrap, wrapped, instance, args, kwargs):
```
```suggestion
async def _awrap(tracer, to_wrap, wrapped, instance, args, kwargs):
```

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/__init__.py:
```diff
@@ -117,6 +166,33 @@ def _set_span_completions(span, response):
             )
 
 
+async def _set_token_usage_a(span, anthropic, request, response):
```
why do you need this function? `count_tokens` isn't an async method

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/__init__.py:
```diff
@@ -178,47 +254,213 @@ def wrapper(wrapped, instance, args, kwargs):
+    # prompt_usage
+    prompt_tokens = 0
+    if kwargs.get("prompt"):
+        prompt_tokens = instance.count_tokens(kwargs.get("prompt"))
```
This call can be costly and increase latency. I'd add a flag for the instrumentation controlling whether this is enabled or not - see the OpenAI instrumentation for an example

nirga commented on packages/opentelemetry-instrumentation-anthropic/tests/test_completion.py:
```diff
@@ -107,12 +107,121 @@ def test_anthropic_multi_modal(exporter):
+            anthropic_span.attributes["llm.prompts.0.user"]
+            == "Tell me a joke about OpenTelemetry"
+    )
+    assert (anthropic_span.attributes.get("llm.completions.0.content"))
```
Can we test that this attribute is actually the response we accumulated in line 209?

nirga reviewed (commented):
Thanks so much for this @AromaticHydrocarbon! I've left a couple of comments. 

Note that you should upload the recordings of your new tests - set an Anthropic Key in your environment, comment out the line with the mock anthropic key in `conftest.py` and then run them locally with `poetry run pytest -v --record-mode=once -vv`

nirga commented:
@AromaticHydrocarbon have you seen my previous comments?

AromaticHydrocarbon commented:
> @AromaticHydrocarbon have you seen my previous comments?

@nirga Thank you very much for your feedback on the code I submitted. I have made modifications and improvements based on your suggestions. 

1. Regarding uploading the recordings of new tests: I apologize for not understanding how GitHub tests are executed. I have now uoload my cassettes.

2. Regarding avoiding code duplication: I have extracted the repeated lines in the stream section and moved the common code into specific methods to reduce duplication.

3. Regarding putting the streaming handling code in a separate file: I have moved the streaming handling code to a separate file.

4.  Regarding the count_tokens function: I carefully reviewed the code and noticed that the count_tokens function in the AsyncAnthropic class is indeed already defined as an asynchronous method. 

5. Regarding adding a flag to control whether to enable costly calls: I referred to the example in the OpenAI instrumentation and used the should_record_stream_token_usage() function in the streaming handling code. 
 
6.Regarding testing whether the accumulated response on line 209 is actually set as an attribute: I have added corresponding content in the tests

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/config.py:
```diff
@@ -0,0 +1,2 @@
+class Config:
+    enrich_token_usage = True
```
A few comment on this Config:
1. I think it's best if its `False` by default. 
2. You also need to have an option to set it during `__init__` of the instrumentation (see the OpenAI instrumentation for an example).
3. You can then pass `True` to it when `traceloop_sdk` initialize it (see how we did it with the OpenAI instrumentation).

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/streaming.py:
```diff
@@ -0,0 +1,143 @@
+        logger.warning("Failed to set completion attributes, error: %s", str(e))
+
+
+def should_record_stream_token_usage():
```
nit: not needed, you can just check the config directly

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/streaming.py:
```diff
@@ -0,0 +1,143 @@
+):
+    complete_response = {"events": [], "model": "", "usage": {}}
+    for item in response:
+        process_response_item(item, complete_response)
```
nit: first yield, then process

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/streaming.py:
```diff
@@ -0,0 +1,143 @@
+):
+    complete_response = {"events": [], "model": ""}
+    async for item in response:
+        process_response_item(item, complete_response)
```
same here - first yield then process

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/streaming.py:
```diff
@@ -0,0 +1,143 @@
+        yield item
+
+    # calculate token usage
+    if should_record_stream_token_usage():
```
this is risky and may fail - better to wrap it with `try` - `except` like we have in `set_completions`

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/utils.py:
```diff
@@ -0,0 +1,15 @@
+from opentelemetry import context as context_api
+
+
+def _set_span_attribute(span, name, value):
```
nit
```suggestion
def set_span_attribute(span, name, value):
```

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/streaming.py:
```diff
@@ -0,0 +1,143 @@
+logger = logging.getLogger(__name__)
+
+
+def process_response_item(item, complete_response):
```
nit
```suggestion
def _process_response_item(item, complete_response):
```

nirga replied:
this is risky and may fail - can we rap it in `try except` and log an error?

nirga commented on packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/streaming.py:
```diff
@@ -0,0 +1,143 @@
+    _set_span_attribute(span, SpanAttributes.LLM_RESPONSE_MODEL, complete_response.get("model"))
+
+
+def _set_completions(span, choices):
```
if this is a copy paste from the `__init__` - better to just re-use it from there (or use this method there). I want to avoid duplications as much as possible cause these methods change often and people tend to forget to look for all the places where a certain attribute is set.

nirga reviewed (commented):
Thanks @AromaticHydrocarbon! I left a couple of more comments. Sorry for the nit-picking, these instrumentations are delicate and we need to make sure we're not breaking anything for our users.

AromaticHydrocarbon commented:
> Thanks @AromaticHydrocarbon! I left a couple of more comments. Sorry for the nit-picking, these instrumentations are delicate and we need to make sure we're not breaking anything for our users.

Thank you for your detailed comments and suggestions. Regarding the issues you raised, I have made the following responses and modifications:

I have completed the corresponding modifications with reference to OpenAI instrumentation, setting enrich_token_usage to False by default and adding a parameter in the __init__ method of AnthropicInstrumentor to control it. 

I have renamed the process_response_item function to _process_response_item.

I have added a try-except block in the _process_response_item function to capture possible exceptions and log errors.

Regarding _set_completions(), I would like to clarify that this part of the code was copied from the streaming handle section of OpenAI instrumentation, not from __init__. 

I have adjusted the order of the code to yield first and then process.

I have renamed the _set_span_attribute function to set_span_attribute.

Thank you again for your feedback. I will continue to work on improving code quality. If you have any other suggestions or opinions, I am very willing to listen and make corresponding adjustments.
